sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UnchangedLocalVariablesShouldBeConst.cs

```diff
@@ -0,0 +1,189 @@
+        {
+            var declaredTypeSyntax = localDeclaration.Declaration?.Type;
+
+            if (declaredTypeSyntax == null)
+            {
+                return DeclarationType.Unknown;
+            }
+
+            var declaredType = semanticModel.GetTypeInfo(declaredTypeSyntax).Type;
+            if (declaredType == null)
+            {
+                return DeclarationType.Unknown;
+            }
+
+            if (declaredType.Is(KnownType.System_String))
+            {
+                return DeclarationType.String;
+            }
+
+            return declaredType.IsValueType
+                ? DeclarationType.Value
+                : DeclarationType.Reference;
+        }
+
+
+        private static bool IsInitializedWithCompatibleConstant(VariableDeclaratorSyntax variableDeclarator,
+            SemanticModel semanticModel, DeclarationType declarationType)
+        {
+            var initializer = variableDeclarator?.Initializer?.Value;
+            if (initializer == null)
+            {
+                return false;
+            }
+            var constantValueContainer = semanticModel.GetConstantValue(initializer);
+            if (!constantValueContainer.HasValue)
+            {
+                return false;
+            }
+
+            var constantValue = constantValueContainer.Value;
+
+            if (constantValue is string)
+            {
+                return declarationType == DeclarationType.String;
+            }
+
+            if (constantValue is ValueType)
+            {
+                return declarationType == DeclarationType.Value;
+            }
+
+            return declarationType == DeclarationType.Reference ||
+                   declarationType == DeclarationType.String;
+        }
+
+        private static bool HasMutableUsagesInMethod(VariableDeclaratorSyntax parameter, ISymbol parameterSymbol,
+            SemanticModel semanticModel)
+        {
+            var methodSyntax = parameter.Parent?.AncestorsAndSelf()?.FirstOrDefault(IsMethodLike);
+            if (methodSyntax == null)
+            {
+                return false;
+            }
+
+            return methodSyntax
+                .DescendantNodes()
+                .OfType<IdentifierNameSyntax>()
+                .Where(MatchesIdentifier)
+                .Any(IsMutatingUse);
+
+            bool IsMethodLike(SyntaxNode arg) =>
+                arg is BaseMethodDeclarationSyntax ||
+                arg is IndexerDeclarationSyntax ||
+                arg is AccessorDeclarationSyntax;
+
+            bool MatchesIdentifier(IdentifierNameSyntax id)
+            {
+                var symbol = semanticModel.GetSymbolInfo(id).Symbol;
+                return Equals(parameterSymbol, symbol);
+            }
+        }
+
+        private static bool IsMutatingUse(IdentifierNameSyntax id)
+        {
+            var parent = id.Parent;
+            if (parent is AssignmentExpressionSyntax assignmentExpression &&
```

Comment:
Does this cover `x += 5` and the variations / * % - | &, etc.? Also shifting operators too...

Reply:
Yes. I'll add more tests.